pkg/workers/send_work.go

```diff
@@ -0,0 +1,263 @@
+package workers
+
+import (
+	"fmt"
+	"time"
+
+	masa "github.com/masa-finance/masa-oracle/pkg"
+	"github.com/masa-finance/masa-oracle/pkg/pubsub"
+	"github.com/masa-finance/masa-oracle/pkg/workers/messages"
+
+	"github.com/asynkron/protoactor-go/actor"
+
+	pubsub2 "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/sirupsen/logrus"
+)
+
+var workerConfig *WorkerConfig
+
+func init() {
+	var err error
+	workerConfig, err = LoadConfig()
+	if err != nil {
+		logrus.Fatalf("Failed to load worker config: %v", err)
+	}
+	workerDoneCh = make(chan *pubsub2.Message, workerConfig.WorkerBufferSize)
+}
+
+type Worker struct {
+	IsLocal  bool
+	NodeData pubsub.NodeData
+	IPAddr   string
+	Node     *masa.OracleNode
+}
+
+func SendWork(node *masa.OracleNode, m *pubsub2.Message) {
+	logrus.Infof("Sending work to node %s", node.Host.ID())
+	props := actor.PropsFromProducer(NewWorker(node))
+	pid := node.ActorEngine.Spawn(props)
+	message := createWorkMessage(m, pid)
+
+	responseCollector := make(chan *pubsub2.Message, 1)
+
+	eligibleWorkers := GetEligibleWorkers(node, message)
+
+	success := tryWorkersRoundRobin(node, eligibleWorkers, message, responseCollector)
+	if !success {
+		logrus.Error("Failed to process the work")
+	}
+}
+
+func tryWorkersRoundRobin(node *masa.OracleNode, workers []Worker, message *messages.Work, responseCollector chan *pubsub2.Message) bool {
+	var localWorker *Worker
+	remoteWorkersAttempted := 0
+
+	logrus.Info("Starting round-robin worker selection")
+
+	// Try remote workers first, up to MaxRemoteWorkers
+	for _, worker := range workers {
+		if !worker.IsLocal {
+			if remoteWorkersAttempted >= workerConfig.MaxRemoteWorkers {
+				logrus.Infof("Reached maximum remote workers (%d), stopping remote worker attempts", workerConfig.MaxRemoteWorkers)
+				break
+			}
+			remoteWorkersAttempted++
+			logrus.Infof("Attempting remote worker %s (attempt %d/%d)", worker.NodeData.PeerId, remoteWorkersAttempted, workerConfig.MaxRemoteWorkers)
+			if tryWorker(node, worker, message, responseCollector) {
+				logrus.Infof("Remote worker %s successfully completed the work", worker.NodeData.PeerId)
+				return true
+			}
+			logrus.Infof("Remote worker %s failed, moving to next worker", worker.NodeData.PeerId)
+		} else {
+			localWorker = &worker
+			logrus.Info("Found local worker, saving for later if needed")
+		}
+	}
+
+	// If remote workers fail or don't exist, try local worker
+	if localWorker != nil {
+		logrus.Info("Attempting local worker")
+		return tryWorker(node, *localWorker, message, responseCollector)
+	}
+
+	// If no workers are available, create a local worker as last resort
+	logrus.Warn("No workers available, creating last resort local worker")
+	lastResortLocalWorker := Worker{
```

Comment:
small nit: mmm wondering here, what's the difference with `localWorker` defined above? isn't `worker.IsLocal` already resolving if the worker is local or not from the list of workers? however I don't know if `GetEligibleWorkers` returns the local node as a worker or not, it seems there is a bit redundancy in the logic.

Reply:
GetEligibleWorkers will return the local worker if it is capable of doing the requested work. You are correct this last check is redundant and should actually never get there. It can be removed.